src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs

```diff
@@ -431,68 +476,91 @@ private void ReadMetadataTableHeader(ref BlobReader memReader, out int[] metadat
+            const ulong validTables = (ulong)(TableMask.V2_0_TablesMask & ~TableMask.PtrTables & ~TableMask.EnCLog & ~TableMask.EnCMap);
 
-            ulong validTables = (ulong)_MetadataTableHeader.ValidTables;
-            int compressedRowCountIndex = 0;
-            for (int i = 0; i < TableIndexExtensions.Count; i++)
+            if ((externalTableMask & ~validTables) != 0)
             {
-                bool fitsSmall;
+                throw new BadImageFormatException(string.Format(MetadataResources.UnknownTables, (TableMask)externalTableMask));
+            }
 
-                if ((validTables & 1UL) != 0)
-                {
-                    int rowCount = compressedRowCounts[compressedRowCountIndex++];
-                    rowCounts[i] = rowCount;
-                    fitsSmall = rowCount < MetadataStreamConstants.LargeTableRowCount;
-                }
-                else
-                {
-                    fitsSmall = true;
-                }
+            externalTableRowCounts = ReadMetadataTableRowCounts(ref reader, externalTableMask);
 
-                referenceSizes[i] = (fitsSmall && !IsMinimalDelta) ? SmallIndexSize : LargeIndexSize;
-                validTables >>= 1;
-            }
+            debugMetadataHeader = new DebugMetadataHeader(
+                MethodDefinitionHandle.FromRowId((int)(entryPointToken & TokenTypeIds.RIDMask)));
+        }
+
+        private const int SmallIndexSize = 2;
+        private const int LargeIndexSize = 4;
+
+        private int GetReferenceSize(int[] rowCounts, TableIndex index)
+        {
+            return (rowCounts[(int)index] < MetadataStreamConstants.LargeTableRowCount && !IsMinimalDelta) ? SmallIndexSize : LargeIndexSize;
+        }
 
+        private void InitializeTableReaders(MemoryBlock metadataTablesMemoryBlock, HeapSizes heapSizes, int[] rowCounts, int[] externalRowCountsOpt)
+        {
+            // Size of reference tags in each table.
             this.TableRowCounts = rowCounts;
 
-            // Compute ref sizes for tables that can have pointer tables for it
-            int fieldRefSize = referenceSizes[(int)TableIndex.FieldPtr] > SmallIndexSize ? LargeIndexSize : referenceSizes[(int)TableIndex.Field];
-            int methodRefSize = referenceSizes[(int)TableIndex.MethodPtr] > SmallIndexSize ? LargeIndexSize : referenceSizes[(int)TableIndex.MethodDef];
-            int paramRefSize = referenceSizes[(int)TableIndex.ParamPtr] > SmallIndexSize ? LargeIndexSize : referenceSizes[(int)TableIndex.Param];
-            int eventRefSize = referenceSizes[(int)TableIndex.EventPtr] > SmallIndexSize ? LargeIndexSize : referenceSizes[(int)TableIndex.Event];
-            int propertyRefSize = referenceSizes[(int)TableIndex.PropertyPtr] > SmallIndexSize ? LargeIndexSize : referenceSizes[(int)TableIndex.Property];
+            // TODO: Shouldn't XxxPtr table be always the same size or smaller than the corresponding Xxx table?
```

Comment:
I'll file a bug. Do we do any testing o Ptr tables at all? The ECMA spec doesn't define them at all. Perhaps we should write a spec for them. I'm not even sure in what scenarios they may show up.